### ISSUE TITLE
Some functionalities are much slower on Linux compared to Windows

### ISSUE THREAD

**Issue by CalinBunea** [dotnet/runtime#48116]
### Description
Hi,
I'm working in BigData field and I have noticed that some functionalities (like String related methods) are extremely slow on .NET Core running on Linux Ubuntu 18.04 compared to Windows Server 2012. The performance difference is like 30X-50X. This issue is present in .NET Core 2.X, 3.X, 5.X. To me this is quite embarrassing and I am seriously thinking to quit working with .NET and start with other language after ~15 years of development in .NET... Is really unreal how Microsoft has released not one but 3 versions of .NET core with this issue and they are not capable of fixing it. I saw a similar complaint from another user and a guy from Microsoft has started to request more information about a very simple String.IndexOf performance complaint. Can't you guys test this out yourself? Is it so hard to write 20 lines of code to test this out? Anyway, I wrote the code for you and I will paste it bellow. I've also noticed that Guid.NewGuid() is also extremely slow on Linux compared to Windows. I have used VMs in Azure for my tests and I have picked exactly the same size for the VMs so the tests should be considered fair.
Just run the code bellow on a Windows OS vs Linux. You can use 10 Million items for test but on Linux it will take forever to execute so you can start with ~2 Million first.
Please spare me from questions like "have you tried with invariant culture or Ordinal compare etc?" Yes, I did tried with same results!

```
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp3._1
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Int32 size = -1;
            if(args.Length>0)
                int.TryParse(args[0],out size);
            if (size < 1)
            {
                Console.WriteLine("Please provide a size for test collection greater than 0");
                return;
            }
            String oneRandomStr = Guid.NewGuid().ToString();
            //generate random data to avoid CPU cache
            String[] randomData = new String[size];
            Int32 idx = -1;
            Stopwatch sw = Stopwatch.StartNew();
            await Task.WhenAll(Enumerable.Range(0, Environment.ProcessorCount).Select(x => Task.Run(() =>
            {
                var currentIdx = Interlocked.Increment(ref idx);
                while (currentIdx < randomData.Length)
                {
                    randomData[currentIdx] = $"{Guid.NewGuid()}{Guid.NewGuid()}";
                    currentIdx = Interlocked.Increment(ref idx);
                }
            })));
            sw.Stop();
            Console.WriteLine($"Random data populated in {sw.Elapsed}");
            sw.Restart();
            foreach (string str in randomData)
            {
                var unusedIdx = str.IndexOf(oneRandomStr, 2);
            }

            sw.Stop();
            Console.WriteLine($"Time spent with String IndexOf :{sw.Elapsed}");
        }
    }
}
```

**Comment by CalinBunea** [comment]
Dude, are you just writing here to get attention or something? How can you possibly ask me what numbers I've got compared to yours when we use different hardware? Just spin a VM on your laptop with Windows or Linux (depending on your host OS) and do a fair comparison using same hardware! Or even better, spin 2 VMs one with Linux Ubuntu 18.04 and one with Windows Server so both VM will suffer any virtualization penalty.
Now, about .NET core implementations...of course Microsoft cannot implement a file open ourside the OS functions, but we are talking here about String.IndexOf which by any means does not have to use any OS functions. Maybe it has to read one time only some information about the OS culture but that should be all. Rest of it should be algorithms and calling the right CPU instructions to gain best performance. Afterall this should be the biggest advantage of JIT - the fact that it should be able to compile IL into machine language using the current CPU architecture and available instructions set.
.NET Core is not some 3rd party library which you can replace or implement it yourself so yes, I do have high expectations from a framework on which I will build my applications and I don't expect to have such silly problems after more than 4 years of development. 
I literally can't trust anymore .NET Core and don't ask me to write my own IndexOf method for my BigData problems.
I've already wrote my own custom collections to handle better huge amount of data (that is another topic which I won't elaborate more here).
What will be next? Should I write my own JIT or what? Maybe I should write the entire .NET Core myself...
Really man, you are so superficial like many people working at Microsoft...

**Comment by terrajobst** [comment]
@CalinBunea 

I have done zero research but I don't doubt that there are benchmarks where one platform performs better than the other. We have seen this in all directions. We generally care about .NET being awesome for workloads we think our customers care about and that very much includes big data processing or string processing in general.

We're quite public about how we approach these things and @perlun described it quite well: we start with scenarios and benchmarks around those and then we see where the bottlenecks are. As @perlun mentioned, the devil tends to be in the details and are often not related to particular algorithms but how our system overall interplays with the idiosyncrasies of the platform we're running on. 

One thing I will say though is that you don't help your case here. You make a bunch of assertions about what we care about, make statements about our technical abilities, and overall question our motives. This is an open source project and as such it has a [code of conduct](https://github.com/dotnet/runtime/blob/master/CODE-OF-CONDUCT.md). Your attitude isn't crossing the line yet, but you're not far off. We're more than willing to work with you on the issues you're raising, but we won't engage with you if your hostile tone doesn't change.

**Comment by tannergooding** [comment]
Here are what I get locally:

<details>
  <summary>Used code, `string.IndexOf(string, int)` - Click to expand!</summary>

```csharp
using System;
using System.Diagnostics;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp3._1
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Int32 size = -1;
            if (args.Length > 0)
                int.TryParse(args[0], out size);
            if (size < 1)
            {
                Console.WriteLine("Please provide a size for test collection greater than 0");
                return;
            }

            Console.WriteLine($"Environment.Is64BitProcess:                 {Environment.Is64BitProcess}");
            Console.WriteLine($"Environment.ProcessorCount:                 {Environment.ProcessorCount}");
            Console.WriteLine($"Environment.OSVersion:                      {Environment.OSVersion}");
            Console.WriteLine($"RuntimeInformation.FrameworkDescription:    {RuntimeInformation.FrameworkDescription}");
            Console.WriteLine($"Size:                                       {size}");

            String oneRandomStr = Guid.NewGuid().ToString();
            //generate random data to avoid CPU cache
            String[] randomData = new String[size];
            Int32 idx = -1;
            Stopwatch sw = Stopwatch.StartNew();
            await Task.WhenAll(Enumerable.Range(0, Environment.ProcessorCount).Select(x => Task.Run(() =>
            {
                var currentIdx = Interlocked.Increment(ref idx);
                while (currentIdx < randomData.Length)
                {
                    randomData[currentIdx] = $"{Guid.NewGuid()}{Guid.NewGuid()}";
                    currentIdx = Interlocked.Increment(ref idx);
                }
            })));
            sw.Stop();
            Console.WriteLine($"Random data populated in {sw.Elapsed}");
            sw.Restart();
            foreach (string str in randomData)
            {
                var unusedIdx = str.IndexOf(oneRandomStr, 2);
            }

            sw.Stop();
            Console.WriteLine($"Time spent with String IndexOf :{sw.Elapsed}");
        }
    }
}
```

</details>

```
Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Microsoft Windows NT 6.2.9200.0
RuntimeInformation.FrameworkDescription:    .NET Core 3.1.11
Size:                                       2000000
Random data populated in 00:00:00.5534036
Time spent with String IndexOf :00:00:00.4083975

Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Unix 4.4.0.19041
RuntimeInformation.FrameworkDescription:    .NET Core 3.1.12
Size:                                       2000000
Random data populated in 00:00:02.7903183
Time spent with String IndexOf :00:00:10.5053876

Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Microsoft Windows NT 10.0.19042.0
RuntimeInformation.FrameworkDescription:    .NET 5.0.2
Random data populated in 00:00:00.5211980
Time spent with String IndexOf :00:00:13.8662883

Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Unix 4.4.0.19041
RuntimeInformation.FrameworkDescription:    .NET 5.0.3
Size:                                       2000000
Random data populated in 00:00:02.8335380
Time spent with String IndexOf :00:00:10.4657736
```

<details>
  <summary>Used code `string.IndexOf(string, int, StringComparison)` - Click to expand!</summary>

```csharp
using System;
using System.Diagnostics;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading;
using System.Threading.Tasks;

namespace ConsoleApp3._1
{
    class Program
    {
        static async Task Main(string[] args)
        {
            Int32 size = -1;
            if (args.Length > 0)
                int.TryParse(args[0], out size);
            if (size < 1)
            {
                Console.WriteLine("Please provide a size for test collection greater than 0");
                return;
            }

            Console.WriteLine($"Environment.Is64BitProcess:                 {Environment.Is64BitProcess}");
            Console.WriteLine($"Environment.ProcessorCount:                 {Environment.ProcessorCount}");
            Console.WriteLine($"Environment.OSVersion:                      {Environment.OSVersion}");
            Console.WriteLine($"RuntimeInformation.FrameworkDescription:    {RuntimeInformation.FrameworkDescription}");
            Console.WriteLine($"Size:                                       {size}");

            String oneRandomStr = Guid.NewGuid().ToString();
            //generate random data to avoid CPU cache
            String[] randomData = new String[size];
            Int32 idx = -1;
            Stopwatch sw = Stopwatch.StartNew();
            await Task.WhenAll(Enumerable.Range(0, Environment.ProcessorCount).Select(x => Task.Run(() =>
            {
                var currentIdx = Interlocked.Increment(ref idx);
                while (currentIdx < randomData.Length)
                {
                    randomData[currentIdx] = $"{Guid.NewGuid()}{Guid.NewGuid()}";
                    currentIdx = Interlocked.Increment(ref idx);
                }
            })));
            sw.Stop();
            Console.WriteLine($"Random data populated in {sw.Elapsed}");
            sw.Restart();
            foreach (string str in randomData)
            {
                var unusedIdx = str.IndexOf(oneRandomStr, 2, StringComparison.Ordinal);
            }

            sw.Stop();
            Console.WriteLine($"Time spent with String IndexOf :{sw.Elapsed}");
        }
    }
}
```

</details>

```
Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Microsoft Windows NT 6.2.9200.0
RuntimeInformation.FrameworkDescription:    .NET Core 3.1.11
Size:                                       2000000
Random data populated in 00:00:00.5428705
Time spent with String IndexOf :00:00:00.1020913

Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Unix 4.4.0.19041
RuntimeInformation.FrameworkDescription:    .NET Core 3.1.12
Size:                                       2000000
Random data populated in 00:00:02.8710434
Time spent with String IndexOf :00:00:00.1353504

Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Microsoft Windows NT 10.0.19042.0
RuntimeInformation.FrameworkDescription:    .NET 5.0.2
Size:                                       2000000
Random data populated in 00:00:00.4901289
Time spent with String IndexOf :00:00:00.1145368

Environment.Is64BitProcess:                 True
Environment.ProcessorCount:                 32
Environment.OSVersion:                      Unix 4.4.0.19041
RuntimeInformation.FrameworkDescription:    .NET 5.0.3
Size:                                       2000000
Random data populated in 00:00:02.8322111
Time spent with String IndexOf :00:00:00.1501700
```

**Comment by tannergooding** [comment]
This was comparing `Windows 10 19042.0` vs `Ubuntu 20.04` running on WSL2 (I also tried on Ubuntu 20.04 proper and got essentially the same results).

The reason that Unix is slower for `string.IndexOf(string, int)` is because it is Culture aware and depends on ICU. This is likewise why this method is slower on .NET 5 than on .NET Core 3.1.
The `string.IndexOf(string, int, StringComparison.Ordinal)` mode is effectively the same speed on both Windows and Linux as its not culture dependent and therefore not dependent on ICU.

CC. @tarekgh, @GrabYourPitchforks

**Comment by stephentoub** [comment]
@tannergooding, you might also try .NET 6 builds, which will include fixes like https://github.com/dotnet/runtime/pull/43065 (I don't think we backported that to 5.0.x, but I'm not 100% positive).

**Comment by tarekgh** [comment]
The GUID performance is already tracked by the issue #13628. And thanks @stephentoub pointing at the optimization we did for linguistic search.

**Comment by SergiiKram** [comment]
Just have done my own tests.

**TLDR:**  performance is different because of different `StringComparison` type applied by default and does not vary with OS platform.

**Long story:**
Tried locally from VS  and from command line - got ~11 seconds (release, self-contained build).
Then tried on a Linux VM on Azure - got ~14 seconds (Ubuntu, B1s size).
Then tried on a Windows VM on Azure - got <1 second (Windows Server 2019 Datacenter, B1s size).
Then explicitly set the comparison type to `StringComparison.Ordinal` - got <1 second on all environments.

**Comment by CalinBunea** [comment]
@SergiiKram, thank you for taking the time to perform benchmarks. Can you also confirm that you have used a random collection of strings, big enough to avoid CPU caching? Also I assume you haven't performed IndexOf on same String object over and over again... 
On my side I was running a very simple task which had to extract one single field value and store its MD5 hash from a file which contains JSON serialized objects on every line. So I was basically did a ReadLine then deserialize and use the field I needed. I was getting about 1.5 million items processed per second on a 8 core Azure VM (I always put performance counters in my code and show them every ~30 seconds)
Since I had to process around 15 billion items I thought to optimize a bit and perform a simple String.IndexOf and Substring to extract my field value. I was very surprised when I saw a speed of only 80.000 items per second. So basically JSON deserialize was 10+ times faster than a basic IndexOf + Substring. Since the code was very simple, I've started to investigate deeper what is causing the performance issue and that is how I found out this problem...

**Comment by SergiiKram** [comment]
@CalinBunea yes, first I used your exact code with 1 000 000 items. Then I modified the code to make sure that there is a substring somewhere in each array item - same results.
I think the key to solution here is to set explicitly `StringComparison.Ordinal` since other variants would give you overhead regardless of platform.
